DIPs/1NNN-DK.md

```diff
@@ -0,0 +1,242 @@
+- `@UserDefinedAttribute`
+
+Moreover, because function attributes are inferred in template functions, forcing them back to the default is not always desirable either.
+Consider:
+
+```
+struct S
+{
+    @nogc nothrow pure @safe:
+
+    void f() { }
+
+    void g() { }
+
+    void toString(O)(ref O sink) if (isOutputRange!(O, char))
+    {
+    }
+}
+```
+
+In this case, the `toString` method should not force the provided Output Range to have an implementation of `sink.put` with all function attributes, but
+neither should `toString` be denied those attributes when it can infer them.
+The only option here is to place all template methods above the `@nogc nothrow pure @safe:` line, or go back to 'attribute soup' by repeating all function attributes for individual members.
+
+What is really desired is something to specify that attributes on `toString` should go back to the default state,
+which this DIP proposes in the form of the `@default` attribute.
+
+## Prior Work
+
+- [DIP 1029 - Add throw as Function Attribute](http://dlang.org/dips/1029)
+The `throw` attribute was added to address the lack of a way to invert the effect of `nothrow:`.
+However, it has not been implemented yet, and can be superseded by this DIP.
+
+- [DIP 1012 - Function Attributes](http://dlang.org/dips/1012)
+Proposes to move function attributes to `core.attribute` enabling traditional methods of attribute manipulation (`AliasSeq`) and introspection.
+This DIP has been abandoned.
+
+## Alternatives
+
+It has been proposed before that attributes could take a boolean parameter evaluated at CTFE (e.g. `@safe(x && y)`), or reset to the default by passing the `default` keyword as a parameter.
+This allows more precision, but it is more complex and verbose:
+
+```D
+@nogc nothrow pure @safe:
+
+// ...
+
+void f(T)(T x) @nogc(default) @nothrow(default) pure(default) @safe(default)
+{
+
+}
+```
+
+Another option is to use `@default` to specify default function attributes for regular functions, but not functions with attribute inference.
+```D
+struct S
+{
+    @default @nogc nothrow pure @safe:
+
+    void f() {}    // attributes apply here
+
+    auto g(T)() {} // does not apply here
+}
+```
+However, this does not provide a way to invert attributes that aren't inferred.
+Neither does it provide a way to invert function attributes on functions without attribute inference.
+
+```
+struct S
+{
+@default @nogc const:
+
+    void usesGc() {} // out of luck, not a template
+
+    void mutates(T)() {} // out of luck, `const` is not inferred
+}
+```
+
+### Syntax
+
+The syntax `@default` is chosen because `default` is already a reserved keyword, but it may not be the clearest name.
+A different name, such as `@reset`, could be considered, despite it potentially being used as a User Defined Attribute already.
+
+The `@` could be omitted when using the `default` keyword, but it is suggested to keep it to make it clear it's related to attributes, and not types or the `default` statement.
```

Comment:
In a declaration context, nothing is nowhere near a statement, including `default` statement in D. A lot of function attributes have no `@` in front of them.

Reply:
I don't get what you're saying. ("nothing is nowhere near a statement"?)

Reply:
> A lot of function attributes have no @ in front of them.

That's the converse of what I'm saying. Things starting with `@` imply attributes, that doesn't mean it holds the other way around as well.

Reply:
The `default` keyword is only used for the `default:` case in a `switch` statement; the spec calls this a [`DefaultStatement`](https://dlang.org/spec/statement.html#DefaultStatement). Function attributes are quite far removed from `switch` statements, there is no possible conflation. (I hope this clarified the confusing sentence.)
Having a keyword after `@` would be quite novel, actually. The @ attributes were introduced as a workaround because the desired keyword (safe, trusted, system, nogc, property, disable, and live) was a legal identifier and not already a keyword, and making it one would have resulted in breakage, and because Walter opposes contextual keywords. With the `@`, it resulted in (almost) no breakage. For `default`, this reasoning does not apply. When `return` became a function attribute, it was not `@return`. I just don’t see why `default` warrants an exception. The position where `default` would appear implies “attribute” according to D syntax.